### ISSUE TITLE
Add Events API/Support for tt_metal for host->CQ and CQ<->CQ synchronization

### ISSUE THREAD

**Issue by kmabeeTT** [tenstorrent/tt-metal#5529]
Tagging @davorchap , @tooniz , @arakhmati , @TT-billteng for visibility of proposal - please take a look, let me know if any concerns. Starting to work on code scoping/implementation in parallel.

==

This "Events" feature been talked about for a while, opening ticket to add some verbosity, and to track design/implementation on tt_metal side which got assigned to me this week. Let me know if I got anything wrong here.

### Intro:

- We are looking for a way to have Host block execution until a HW CQ makes progress to a certain command, a kind of "finer grain" barrier, rather than coarse barrier (to wait until CQ is fully drained), like today.
- Similiarly, looking for the same feature between multiple HW CQ's. As in, a way to have dependencies enforced between commands for when there are multiple HW CQ's used by a graph, such that say CQ2 can block execution until CQ1 has made progress to a certain command.

I stole a picture from @arakhmati and added words here to illustrate that case:

![Screenshot 2024-02-21 at 12 01 20 PM](https://github.com/tenstorrent-metal/tt-metal/assets/118925087/1bfbfed6-38ff-45ea-aaf4-4d0bd7e274af)

To start, @TT-billteng just merged change to add initial `Event` struct to `tt_metal/impl/event/event.hpp` as :

```
namespace tt::tt_metal
{
    class Device;
    struct Event
    {
        Device * device;
        uint32_t cq_id;
        uint32_t event_id;
    };
}
```

.... and these APIs to `tt-metal/tt_eager/queue/queue.hpp` as stubs for now (until changes for this ticket implemented in tt_metal) - comments added by me here:

```
void QueueRecordEvent(CommandQueue&, Event&);  // Non-blocking populate Event struct and record it for a CQ
void QueueWaitForEvent(CommandQueue&, Event&); // Non-blocking CQ will wait for an event completed on another CQ
void EventSynchronize(Event&);                 // Blocking - host will wait until event completed on a CQ
void QueueSynchronize(CommandQueue&);          // Blocking - host will wait until CQ entirely drained
```
The expectation here is that empty Event struct is created by user, passed by reference and filled-in by `QueueRecordEvent()` and then used by `QueueWaitForEvent()` ie:

```
Event math_event;
QueueRecordEvent(cq1, math_event);
QueueWaitForEvent(cq2, math_event);
```

### Design / Proposal

ttnn change:

1. There was suggestion in meeting from @davorchap last week to rename `QueueRecordEvent()` and `QueueWaitForEvent()` to be `EnqueueQueueRecordEvent()` and `EnqueueQueueWaitForEvent()` for consistency with other functions to denote they are non-blocking and issued to device.  It was suggested to bring this up with Moreh before changing.

tt_metal spec:

A couple choices here, but since ttnn/eager API has already been decided with Moreh, I think we better make tt_metal API very similiar, so:

1. Create the following tt_metal API's that 1:1 map with ttnn APIs (these have `Enqueue` prefix, am anticipating the same can be made for ttnn API):

```
void EnqueueQueueRecordEvent(CommandQueue&, Event&); 
void EnqueueQueueWaitForEvent(CommandQueue&, Event&);
void EventSynchronize(Event&);
void QueueSynchronize(CommandQueue&);
```

Usage - mapping to 4-op example image above:

```
// Create vector of 2 empty events, to be populated by Record(). 
// Or use named structs like exp_add_sync_event, add_gelu_sync_event.
std::vector<Event> events(2); 

EnqueueProgram(cq1, sqrt_program, blocking=False);
EnqueueProgram(cq1, exp_program, blocking=False);

// Setup the synchronization. CQ2 must wait until CQ1 exp finished
EnqueueQueueRecordEvent(cq1, events[0]); // Will populate event struct
EnqueueQueueWaitForEvent(cq2, events[0]);

EnqueueProgram(cq2, add_program, blocking=False);

// Setup the synchronization. CQ1 must wait until CQ2 add finished
EnqueueQueueRecordEvent(cq2, events[1]); // Will populate event struct
EnqueueQueueWaitForEvent(cq1, events[1]);

EnqueueProgram(cq1, gelu_program, blocking=False);
```

2. Briefly considered other ideas like `EnqueueQueueRecordEvent(cq)` returning an Event object to user (so they don't have to create empty Event objects) but it sounds like current API is preferred and already agreed upon from ttnn level (this would clash), so not considering this right now.

3. Ignore 32-bit event_id wrapping back to 0, to start. From talking to Andrew, we don't think it will happen anytime soon, can assert if event_id is seen to wrap.  If we do need to support it, let me know and we can spec out how to handle it.

### Implementation:

Currently TBD/WIP -  will start working on this now and scoping out code, will fill in more details once I know them.

1. `EnqueueQueueRecordEvent()` will populate Event struct with latest event_id from auto-incrementing event_id already used for commands, so this feature "just works" by waiting for completed event_id to be greater than waited-for event_id. 

2. Host->Device sync via `EventSynchronize()`: sounds like the expectation here from Davor in meeting last week is that Device would write event_id to completion queue in Host Hugepage (already does this today for cmds), and the support in `EventSynchronize(Event)` will be real simple to have Host wait for event_id completed for a CQ to be greater than that waited (`Event.id`) - this is Host reading Host memory.

3. For Device->Device (CQ to CQ sync) via `EnqueueQueueWaitForEvent()`: the same completion queue event_id reading can be done, but it would be Device reading Host completion queue for a given CQ to wait for completed event_id greater than waited-for event_id, and spinning with a tiny sleep of say 10us (to avoid non-stop device->host reads) until that occurs. 

4. As a performance optimization, we can consider eliminating Device->Host reading above, and have CQ1 read CQ2 "last completed event_id" or similiar, stored on device directly.

**Comment by kmabeeTT** [comment]
Hey @TT-billteng, how was `cq_id` on `Event` struct intended to be used to access a `CommandQueue` object from `EventSynchronize(Event)` path?  

I got `EnqueueQueueRecordEvent(cq, Event)` implemented in tt_metal, but realize now that since `EventSynchronize(Event)` does not take a `CommandQueue` object as an argument, and as far as I can tell there isn't any global map of CQ's by ID laying around, I don't see how the `cq_id` field can be used to fine the right `CommandQueue` object to inspect it's last completed event... feels like a cq argument is needed like all the other functions unless I'm missing something (hopefully I am since I agree it feels ~redundant).

**Comment by kmabeeTT** [comment]
Small update - 

> as far as I can tell there isn't any global map of CQ's by ID laying around (at least not in tt_metal), I don't see how the `cq_id` field can be used to find the right `CommandQueue` object to inspect it's last completed event... 

Maybe I'm wrong and prev idea isn't needed, I was looking around more, and came across above code, and now I see there is a way to go from `device` ptr to `HWCommandQueue` object which can be looked up by `cq_id`. Maybe this will solve my concerns... need to look more later.

```
HWCommandQueue& CommandQueue::hw_command_queue() {
    return this->device()->hw_command_queue(this->cq_id);
}
```

**Comment by razorback3** [comment]
Looked through all the discussions above.
First of all, the current implementation looks fine.

Moreover, Moreh does not take care much about the API's name whether it is "EnqueueQueue~" or "Queue~".

Also, the alternative way of implementation is also good.
> Briefly considered other ideas like EnqueueQueueRecordEvent(cq) returning an Event object to user (so they don't have to create empty Event objects) but it sounds like current API is preferred and already agreed upon from ttnn level (this would clash), so not considering this right now.

What we only focus on is a functionality to synchronize between two queues with events.

Thank you.

**Comment by kmabeeTT** [comment]
The 4th API function I proposed in ticket, `QueueSynchronize(cq) `seems not necessary at tt_metal, now that I look closer and tidying up changes, it's the exact same thing as `Finish(cq)` which is already exposed in `host_api.cpp` - so I'm going to not include that in these changes.  I see ttnn implementation is not a stub here, and they already call Finish:

```
void QueueSynchronize(CommandQueue& q) { Finish(q); }
```
If anyone objects let me know.

**Comment by kmabeeTT** [comment]
I got initial implementation working and cleaned up, and in process of review/ci/merge - just wanted to flag something multi-cq related to Moreh folks @razorback3 that came up in discussion that is ~orthogonal to Events:

1. There is an existing restriction that for multi-cq (`num_hw_cqs`>1 on `Device` constructor) , programs are only allowed to be enqueued on `cq_id=0` (there are asserts that guard against this in `EnqueueProgram()`.. something to be mindful of. This restriction doesn't apply to data (Reads/Writes, and now Events), they can be inserted on say `cd_id=1` when `num_hw_cqs=2` for example.  I include some gtests that do this for the new Event APIs (`EnqueueQueueRecordEvent()` and `EnqueueQueueWaitForEvent()` 

Maybe this is already known. I'm checking with Andrew to see if there is some plan to remove this restriction.

**Comment by kmabeeTT** [comment]
Update here - initial changes are merged at 4ec8d35b9 as discussed above.  Keeping this ticket open since needs to be revisted in coming days for:

1. testing with async queues once changes for them are released (there is chance that it will "just work")
2. multichip once multichip fast-dispatch support and other cross-chip communication API changes made (currently not supported with Events, assert added to ensure record event and wait-for-event are from same device) in `command_queue.cpp`:

```
    // #5529 - Cross chip sync needs to be implemented. Currently, we only support sync on the same chip.
    TT_ASSERT(this->sync_event.device == this->device,
            "EnqueueWaitForEvent() cross-chip sync not yet supported. Sync event device: {} this device: {}",
            this->sync_event.device->id(), this->device->id());
```

**Comment by davorchap** [comment]
Looks amazing, thank you @kmabeeTT !

Perhaps we could shorten to EnqueueRecordEvent and EnqueueWaitForEvent? 

"EnqeueQueueXYZ" read a bit funny.

**Comment by kmabeeTT** [comment]
> Perhaps we could shorten to EnqueueRecordEvent and EnqueueWaitForEvent?

Sure, no prob - made this changes (global search and replace, update APIs, functions, comments, files, docs etc) in PR https://github.com/tenstorrent-metal/tt-metal/pull/5837, will merge soon. FYI @razorback3 for tiny name change.

**Comment by kmabeeTT** [comment]
Merged #5837 for the name change (EnqueueQueue_* -> Enqueue_*) at d978fdab7427ddd0564948dc82cd29a3b9d0c045.

**Comment by kmabeeTT** [comment]
FYI @razorback3 while this is still fresh and unused, we would like to change these newly added APIs to take `std::shared_ptr<Event>` instead of `Event&` - do you have any concerns with this change?


```
void EnqueueRecordEvent(CommandQueue& cq, Event &event);
void EnqueueWaitForEvent(CommandQueue& cq, Event &event);
void EventSynchronize(Event &event);
```

Becomes:

```
void EnqueueRecordEvent(CommandQueue& cq, std::shared_ptr<Event> event);
void EnqueueWaitForEvent(CommandQueue& cq, std::shared_ptr<Event> event);
void EventSynchronize(std::shared_ptr<Event> event);
```

The motivation here is to better protect against unsafe usage when "async queues" are enabled (upcoming changes  from https://github.com/tenstorrent-metal/tt-metal/issues/5804 and using `TT_METAL_CQ_ASYNC_MODE=1`) where is it easier for user to fumble and have Event refs go out of scope now in main thread now that `Enqueue_*` operations host code are dispatched to a child thread, causing corruption in the child thread that can lead to various undefined behavior.  Simplest tests like this are no longer safe unless it were a shared_ptr.

```
// Failure. Original test, this does not work with async queues. event goes out of scope in this thread before used by async thread.
TEST_F(CommandQueueFixture, TestEventsEnqueueRecordEventIssueQueueWrapExample0) {
    constexpr size_t num_events = 100000;
    for (size_t i = 0; i < num_events; i++) {
        Event event; // <-- goes out of scope.
        EnqueueRecordEvent(this->device_->command_queue(), event);
        EXPECT_EQ(event.event_id, i);
    }
    Finish(this->device_->command_queue());
}```

I made local changes to convert API to use `shared_ptr<Event>` and issue went away, plan to rebase, and open PR and merge in next couple days unless any objections.

**Comment by kmabeeTT** [comment]
> No problem. Nice work :)

Thanks for quick reply! Merged at 547798575 just now the change to use `shard_ptr<Event>` instead of `Event&`. 

Related to an earlier bullet on async testing WIP, I found an issue when local testing with upcoming async-cq changes (`EnqueueRecordEvent()` gets incorrect event_id because it's value is picked from main thread, yet other cmds are picked from async worker threads, can cause event_id to be out-of-order; not good), will be working on fix tonight/tomorrow there, and further async+events testing.

**Comment by kmabeeTT** [comment]
> Related to an earlier bullet on async testing WIP, I found an issue when local testing with upcoming async-cq changes (`EnqueueRecordEvent()` gets incorrect event_id because it's value is picked from main thread, yet other cmds are picked from async worker threads, can cause event_id to be out-of-order; not good), will be working on fix tonight/tomorrow there, and further async+events testing.

This is fixed at 553027e6e and Events tests updated for async cq at 9dca57876.  Initial async cq changes were merged earlier today too.  Going to close this ticket now.

**Comment by kmabeeTT** [comment]
Screnshot from slide I shared today. @razorback3 @sangwon-chae 

![image](https://github.com/tenstorrent-metal/tt-metal/assets/118925087/f0b88afa-0d1b-4678-b23a-92630b76d5bd)

**Comment by razorback3** [comment]
@kmabeeTT 
This issue has already been resolved, but I would like to ask you for one more event-related API.
Would it be possible to add a 'QueryEvent' API with the same functionality as [cudaEventQuery](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EVENT.html#group__CUDART__EVENT_1g2bf738909b4a059023537eaa29d8a5b7)?
It would be great if you could let me know whether it is possible and if it doesn't require much work.

**Comment by kmabeeTT** [comment]
> Would it be possible to add a 'QueryEvent' API with the same functionality as [cudaEventQuery](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EVENT.html#group__CUDART__EVENT_1g2bf738909b4a059023537eaa29d8a5b7)? 

Hey @razorback3  - thanks for suggestion.. I took a look, this sounds cool and simple and is basically a derivative of `EventSynchronize(event)` on our side (instead of waiting, just return if it's ready).  We don't have any enumerated return status codes yet anywhere, so if we're okay with returning bool, then I can implement this easily in main (Fast Dispatch 1.x) and FD2 branches right away. 

```
bool EventQuery(std::shared_ptr<Event> event); // true if event is completed, false if event is not completed.
```

Sound good? Re-opening.

**Comment by razorback3** [comment]
> > Would it be possible to add a 'QueryEvent' API with the same functionality as [cudaEventQuery](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EVENT.html#group__CUDART__EVENT_1g2bf738909b4a059023537eaa29d8a5b7)?
> 
> Hey @razorback3 - thanks for suggestion.. I took a look, this sounds cool and simple and is basically a derivative of `EventSynchronize(event)` on our side (instead of waiting, just return if it's ready). We don't have any enumerated return status codes yet anywhere, so if we're okay with returning bool, then I can implement this easily in main (Fast Dispatch 1.x) and FD2 branches right away.
> 
> ```
> bool EventQuery(std::shared_ptr<Event> event); // true if event is completed, false if event is not completed.
> ```
> 
> Sound good? Re-opening.

Perfect 😄
Thanks for your fast response.

**Comment by kmabeeTT** [comment]
> Perfect 😄 Thanks for your fast response.

No prob - this is done and merged to tt-metal main at 06828e32f and no updates needed for FD2, just works.
